usbc: add checks for results of TCPC related functions

barnas-michal opened the pull request:
This PR adds checks for the failing paths in the USB-C PD stack. It adds option for the tcpc_init function to return error or postpone value. Error causes the type-c layer to be disabled, where postpone value means that the TCPC may need more time (for powering up) and needs more tries.
Other failures in the stack means that there is a communication problem with the TCPC, so the layer is going into error recovery state.

barnas-michal committed:
usbc: change type for functions that may return error code

TCPC API functions should be able to inform the caller if the function
is supported and successfully executed. Bool values doesn't allow this
so it is needed to change the type to int.
For is_rx_pending_msg function the return code should conform to
existing error codes, so in case of function being not supported,
the -ENOSYS should be returned. In case of successful execution,
if there is no pending message, the -ENODATA should be returned and
in case of message pending, the value of 0.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add check if tcpc initialization was successful

The TCPC driver initialization can be unsuccessful and it should be
checked and logged if that's the case.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: check if message is pending before trying to receive it

If TCPC supports checking for pending messages, it should be done
instead of forcefully trying to read the message and getting no-data
result. This helps with powereficiency for chips that go to sleep
when there is no message pending. Trying to read the message can
wake up the chip providing higher power usage.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: check if sampling CC lines were successful

When the get_cc function returns error, the CC lines should be
considered as open. It may happen either due to some chip malfunction
or chip going into sleep mode and needing time to reinitialize.
TCPC should only go to sleep if there is no partner attaches, so
in case of this error, the lines should be considered as disconnected.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: check if setting the CC and VCONN was successful

This commit adds check in cc_open entry state and logs if setting
either of the values were unsuccessful.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: merge the is_rx and get_rx_pending_msg functions

These two functions are used together so there is no need for
splitting them into two functions. This commit also makes this
function required to be implemented by the TCPC driver.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add checks for results of TCPC functions

This commit adds checks for results and in case of error, handles it.
The failed initialization of the TCPC can postpone initialization in
case the TCPC needs more time after powering up. Otherwise if other
error happened, the Type-C Layer will be disabled.
Errors in other places of the USB-C stack, like communication errors,
moves the state machine to the error recovery, restarting the
communication with parner.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

keith-zephyr commented on drivers/usb_c/tcpc/ucpd_stm32.c:
Typos in the 1st commit message:
"should conforms" -> "should conform"
"functino" -> "function"

keith-zephyr commented on drivers/usb_c/tcpc/ucpd_stm32.c:
```diff
@@ -1027,21 +1027,20 @@ static int ucpd_transmit_data(const struct device *dev,
  * @brief Tests if a received Power Delivery message is pending
  *
- * @retval true if message is pending, else false
+ * @retval 0 if message is pending, else -ENODATA
```
Searching for other APIs that return `int`, but have `is` in the name, I found a few examples that return 0 or 1 for non-errors. 
[gpio_pin_is_input](https://github.com/zephyrproject-rtos/zephyr/blob/d885048637d6070182cf3ca832631dc91c85ad14/include/zephyr/drivers/gpio.h#L1080)
[rtc_alarm_is_pending](https://github.com/zephyrproject-rtos/zephyr/blob/d885048637d6070182cf3ca832631dc91c85ad14/include/zephyr/drivers/rtc.h#L344)

I prefer this approach so that negative return values are only used for errors.

barnas-michal replied:
I was talking with @sambhurst to find a best approach to this function and we found the ENODATA error code so we thought about using it. However, in the next commit this function is merged with the receive data one, so the behavior and return codes are different.
But anyway, thanks for sharing the functions you found, I will remember to use that approach.

sambhurst replied:
Thanks Keith, I like this approach too.

jfischer-no commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -118,7 +123,11 @@ static void tc_init(const struct device *dev)
 	/* Initialize the TCPC */
-	tcpc_init(tcpc);
+	ret = tcpc_init(tcpc);
+	if (ret < 0) {
```
Knowing the context why you write it this way and not `if (ret != 0)` or `if (ret)`, you introduce undefined behavior because positive return value for `tcpc_init();` is not defined and only 0 is defined as success.

barnas-michal replied:
Hmm, it's really nice point here. To be honest, I used the comparison "< 0" because I found many functions that also had a negative error codes and 0 on success and their result were checked that way. I will change that to be more defensive

barnas-michal replied:
Done

barnas-michal committed:
usbc: improve error handling in the ucpd_stm32 and vbus_adc

As noted in PR#63165 checking of result should use comparison to
success value instead of checking if result is negative. It will
allow to check if function returned invalid but positive value.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

keith-zephyr commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -1175,9 +1175,11 @@ static void prl_rx_wait_for_phy_message(const struct device *dev)
 	/* Get the message */
-	if (tcpc_receive_data(tcpc, rx_emsg) <= 0) {
+	pending_rx = tcpc_is_rx_pending_msg(tcpc, NULL);
+	if (pending_rx == -ENODATA || tcpc_receive_data(tcpc, rx_emsg) <= 0) {
```
Even though you're deleting `tcpc_is_rx_pending_msg()` in another commit, the logic here isn't correct and would break bisectability.

```suggestion
	if (pending_rx == 0 || tcpc_receive_data(tcpc, rx_emsg) <= 0) {
```

barnas-michal replied:
Right, missed this one, thanks!

keith-zephyr commented on drivers/usb_c/tcpc/ucpd_stm32.c:
```diff
@@ -1027,10 +1027,10 @@ static int ucpd_transmit_data(const struct device *dev,
+static int ucpd_is_rx_pending_msg(const struct device *dev, enum pd_packet_type *type)
 {
 	struct tcpc_data *data = dev->data;
 	bool ret;
```
Non-blocking comment, this should be 
```suggestion
	bool int;
```
but you're deleting this function completely in a separate commit.

barnas-michal replied:
Fixed that, thanks!

keith-zephyr commented on drivers/usb_c/tcpc/ucpd_stm32.c:
```diff
@@ -1024,45 +1024,25 @@ static int ucpd_transmit_data(const struct device *dev,
-	/* Make sure we have a message to retrieve */
-	if (!ucpd_is_rx_pending_msg(dev, NULL)) {
-		return -EIO;
+	if (!msg) {
```
Per the misra r.14.4, you need explicitly check against NULL here.

https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html
https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_14_04.c

At one point there was  a compliance check for this, but it got disabled.

barnas-michal replied:
Done

keith-zephyr commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -55,8 +56,15 @@ void tc_run(const struct device *dev, const int32_t dpm_request)
+			}
+
 			break;
 		}
```
```suggestion
		if (ret != -EAGAIN) {
			/* Transition to Disabled State */
			LOG_ERR("Disabling the Type-C Layer");
			data->tc_enabled = false;
			tc_set_state(dev, TC_DISABLED_STATE);
		}

		if (ret != 0) {
			break;
		}
```

barnas-michal replied:
It will require double checking for `!= 0`, but I did that for you to compare

keith-zephyr commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -131,17 +139,25 @@ static int tc_init(const struct device *dev)
 	/* Initialize the TCPC */
 	ret = tcpc_init(tcpc);
-	if (ret < 0) {
+	if (ret) {
```
Please update all `if ret` to `if ret != 0` throughout.
